src/library/attribute_manager.h

```diff
@@ -7,65 +7,149 @@ Author: Leonardo de Moura
 #pragma once
 #include "kernel/environment.h"
 #include "library/io_state.h"
+#include "util/priority_queue.h"
 
 #ifndef LEAN_DEFAULT_PRIORITY
 #define LEAN_DEFAULT_PRIORITY 1000u
 #endif
 
 namespace lean {
+
+struct attr_data {
+    virtual unsigned hash() const {
+        return 0;
+    }
+};
+
+typedef std::shared_ptr<attr_data> attr_data_ptr;
+struct attr_config;
+
+class attribute {
+    friend struct attr_config;
+private:
+    std::string m_id;
+    std::string m_descr;
+    std::string m_token;
+protected:
+    environment set_core(environment const &, io_state const &, name const &, attr_data_ptr, bool) const;
+    virtual void write_entry(serializer &, attr_data const &) const = 0;
+    virtual std::unique_ptr<attr_data> read_entry(deserializer &) const = 0;
+public:
+    attribute(char const * id, char const * descr) : m_id(id), m_descr(descr), m_token(std::string("[") + id) {}
+    std::string const & get_name() const { return m_id; }
+    std::string const & get_token() const { return m_token; }
+};
+
+typedef std::shared_ptr<attribute const> attribute_ptr;
+
+/** \brief An attribute without associated data or priority declaration */
+class basic_attribute : public attribute {
+protected:
+    virtual void write_entry(serializer &, attr_data const &) const override final {}
+    virtual std::unique_ptr<attr_data> read_entry(deserializer &) const override final { return {}; }
+public:
+    basic_attribute(char const * id, char const * descr) : attribute(id, descr) {}
+    virtual environment set(environment const & env, io_state const & ios, name const & n, bool persistent) const {
+        return set_core(env, ios, n, attr_data_ptr(new attr_data), persistent);
+    }
+};
+
+// TODO(sullrich): Let the attribute handle priority parsing itself by introducing a custom
+// syntax (something like [attr:prio]?)
+class prio_attribute : public attribute {
+protected:
+    virtual void write_entry(serializer &, attr_data const &) const override final {}
+    virtual std::unique_ptr<attr_data> read_entry(deserializer &) const override final { return {}; }
+public:
+    prio_attribute(char const * id, char const * descr) : attribute(id, descr) {}
+    virtual environment set(environment const & env, io_state const & ios, name const & n, unsigned prio,
+                            bool persistent) const;
+};
+
+template<typename Data>
+class typed_attribute : public attribute {
```

Comment:
For `typed_attribute`, the data class is now responsible for serialization, which reduces `unsigned_params_attribute` to a typedef. But `basic_attribute` and `prio_attribute` still aren't `typed_attribute`s because they have different `on_set` signatures.